dm/pkg/storage/utils_test.go

```diff
@@ -0,0 +1,344 @@
+		"s3://bucket4/prefix/path?access-key=NXN7IPIOSAAKDEEOLMAF&secret-access-key=nREY/7Dt+PaIbYKrKlEEMMF/ExCiJEX=XMLPUANw",
+		"s3://bucket4/prefix/path.mysql-replica-01?access-key=NXN7IPIOSAAKDEEOLMAF&secret-access-key=nREY/7Dt+PaIbYKrKlEEMMF/ExCiJEX=XMLPUANw",
+		"s3://bucket4/prefix/path.t-Ë!s`t?access-key=NXN7IPIOSAAKDEEOLMAF&secret-access-key=nREY/7Dt+PaIbYKrKlEEMMF/ExCiJEX=XMLPUANw",
+	}
+
+	testIsS3Results := []bool{
+		false, false, false, false, false, false, false, false, true, true, true, true, true, true,
+	}
+
+	for i, testPath := range testPaths {
+		isS3 := IsS3Path(testPath)
+		require.Equal(t, testIsS3Results[i], isS3)
+	}
+
+	testAdjustResults := []struct {
+		hasErr bool
+		res    string
+	}{
+		{false, ""},
+		{true, "parse (.*)1invalid:(.*): first path segment in URL cannot contain colon*"},
+		{false, "file:///tmp/storage.mysql-replica-01"},
+		{false, "/tmp/storage.mysql-replica-01"},
+		{false, "./tmp/storage.mysql-replica-01"},
+		{false, "./tmp/storage.mysql-replica-01"},
+		{false, "./tmp/storage.t-Ë!s`t.mysql-replica-01"},
+		{false, "tmp/storage.mysql-replica-01"},
+		{false, "s3:///bucket/more/prefix.mysql-replica-01"},
+		{false, "s3://bucket2/prefix.mysql-replica-01"},
+		{false, "s3://bucket3/prefix/path.mysql-replica-01?endpoint=https://127.0.0.1:9000&force_path_style=0&SSE=aws:kms&sse-kms-key-id=TestKey&xyz=abc"},
+		{false, "s3://bucket4/prefix/path.mysql-replica-01?access-key=NXN7IPIOSAAKDEEOLMAF&secret-access-key=nREY/7Dt+PaIbYKrKlEEMMF/ExCiJEX=XMLPUANw"},
+		{false, "s3://bucket4/prefix/path.mysql-replica-01?access-key=NXN7IPIOSAAKDEEOLMAF&secret-access-key=nREY/7Dt+PaIbYKrKlEEMMF/ExCiJEX=XMLPUANw"},
+		{false, "s3://bucket4/prefix/path.t-Ë!s`t.mysql-replica-01?access-key=NXN7IPIOSAAKDEEOLMAF&secret-access-key=nREY/7Dt+PaIbYKrKlEEMMF/ExCiJEX=XMLPUANw"},
+	}
+
+	for i, testPath := range testPaths {
+		newDir, err := AdjustPath(testPath, "mysql-replica-01")
+		if testAdjustResults[i].hasErr {
+			require.Error(t, err)
+			require.Regexp(t, testAdjustResults[i].res, err.Error())
+		} else {
+			require.NoError(t, err)
+			require.Equal(t, testAdjustResults[i].res, newDir)
+		}
+	}
+
+	// special suffix
+	testAdjustResults = []struct {
+		hasErr bool
+		res    string
+	}{
+		{false, ""},
+		{true, "parse (.*)1invalid:(.*): first path segment in URL cannot contain colon*"},
+		{false, "file:///tmp/storage.t-Ë!s`t"},
+		{false, "/tmp/storage.t-Ë!s`t"},
+		{false, "./tmp/storage.t-Ë!s`t"},
+		{false, "./tmp/storage.mysql-replica-01.t-Ë!s`t"},
+		{false, "./tmp/storage.t-Ë!s`t"},
+		{false, "tmp/storage.t-Ë!s`t"},
+		{false, "s3:///bucket/more/prefix.t-Ë!s`t"},
+		{false, "s3://bucket2/prefix.t-Ë!s`t"},
+		{false, "s3://bucket3/prefix/path.t-Ë!s`t?endpoint=https://127.0.0.1:9000&force_path_style=0&SSE=aws:kms&sse-kms-key-id=TestKey&xyz=abc"},
+		{false, "s3://bucket4/prefix/path.t-Ë!s`t?access-key=NXN7IPIOSAAKDEEOLMAF&secret-access-key=nREY/7Dt+PaIbYKrKlEEMMF/ExCiJEX=XMLPUANw"},
```

Comment:
should the special characters be URL encoded (use Escape not Unescape)?

Reply:
after some test about special characters, i handle separately whether it is a url format, and url format will be escaped for safety. bwt, i add some unit test and integrate test.